Performance improvements & speedtest clients

jagerman opened the pull request:
This PR adds a speedtest client/server to diagnose performance and, based on that, includes significant performance improvements in how we send/receive packets.

See individual commits (especially `Packet batching, sendmmsg, GSO, performance`) for the details.

## Known issue

Network close is broken by this PR for two reasons:
- During Network::close() we do a walk to close handles, but uvw walk assumes it owns *all* libuv types with data elements, and so it blindly casts our uv_udp_t to a uvw::udp_handle, joy!
- Our uv_udp_t destructors *should* be closing on destruction, but aren't (I think because our shared_ptr udp's are shared in too many places and so the destructors aren't firing).

These are not going to be fixed here; we will instead refactor endpoint to de-tangle the socket ownership and management.

jagerman committed:
Add speedtest

Add speedtest client and server that send data to test performance.

It supports multi-stream sending, various data sizes, data verification
on the server.

TODOs:

- implement reverse mode (i.e. server sends data to client)

dr7ana commented on include/quic/connection.hpp:
```diff
@@ -24,7 +24,7 @@ namespace oxen::quic
     class Client;
     class Handler;
 
-    class Connection : public std::enable_shared_from_this<Connection>
```
@jagerman what's the motivation behind removing this? i was thinking it may be useful at some point to hold a weak pointer to the connection in `Stream.user_data`

jagerman replied:
No reason we can't re-add it later if we want it, but for now it isn't in use.

dr7ana commented on include/quic/connection.hpp:
```diff
@@ -92,9 +92,15 @@ namespace oxen::quic
 
+        /// Returns a pointer to the owning Server, if this is a Server connection, nullptr
+        /// otherwise.
         Server* server();
+        const Server* server() const;
 
+        /// Returns a pointer to the owning Client, if this is a Client connection, nullptr
+        /// otherwise.
         Client* client();
+        const Client* client() const;
```
would it be better to do an operator overload for the client and server classes s.t. we can pass them as const?

jagerman committed:
Packet batching, sendmmsg, GSO, performance

This adds new compile-time modes for sending packets: the current libuv
send, which is fairly CPU heavy (as each packet requires extra
allocation and gets shoved into a queued), libuv try-send, sendmmsg, and
sendmmsg+GSO.

This allows us to send up to 24 packets in a single syscall (with the
latter two), which noticeably reduces CPU usage when the pipes are full,
for example when transmitting on a local network or to localhost.

When not using libuv queuing directly, only if sending immediately would
block do we fall back on libuv's send queuing to deliver the packets as
soon as the socket becomes writeable (and re-wake ourselves to continue
flushing streams).

The compile-time mode is controlled via a cmake flag,
-DLIBQUIC_SEND=mode where mode is one of `libuv_queue`, `libuv_try`,
`sendmmsg`, or `gso`.  The latter two are only supported on FreeBSD and
Linux, and the last only on Linux.

Closely related to this, we switch on libuv's recvmmsg support (by
default, though again only does something in FreeBSD/Linux).  uvw,
however, is entirely broken when this flag is switched on, so this
commit drops uvw's management of the udp handle and switches to plain
libuv.

This switch (uvw->libuv) also improved throughput by around 10% since we
can avoid the extra junk and extra allocations that uvw forces on us.

As part of this deep dive into the internals various other improvements
were made that, jointly, vastly improve performance:

- various hot path log statements were downgraded to trace so that they
  don't impact release builds

- flush_streams logic got completely rewritten to use a single loop and
  stack rather than two separate loops (with a nested loop inside one of
  those).  This greatly reduces code duplication.

- flush_streams now randomizes the starting stream; previously we were
  always starting from the first one and going forward, but that biased
  our selection to favour earlier streams since we can only sent
  `max_stream_packets` at once; this both gives earlier streams more
  potential packets, and could entirely flush later streams (e.g. if we
  have max 45 stream packets and 50 streams, the last 5 streams would
  not be able to send *any* packets if all of 1-45 want to send.

  By randomizing the stream we start on each time we enter
  flush_streams, this can still happen for *one* flush_streams call, but
  we spread out stream selection across flush_streams calls.

- Use much larger window sizes (using the values from ngtcp2 example
  defaults); we were starving on high latency, high-throughput stream
  transmission.

- Removed ClientContext/ServerContext destructors: they were closing
  handles, but that isn't necessarily correct (because handles can be
  shared), and is now instead done in the destructor we give when we
  create the shared_ptr<uv_udp_t>.

- Address and Path got completely rewritten to be uvw-free: they now
  work with a sockaddr which can be either a sockaddr_in or
  sockaddr_in6.

- Working IPv6 support.  (This came for free with the previous change).

- Address and Path are now fmt formattable: an Address formats as either
  `1.2.3.4:567` (IPv4) or `[a:b::1]:789` (IPv6).  Paths format as
  `{LOCALADDR ➙ REMOTEADDR}`.

- Removed universal_handle: it didn't work properly to share the UDP
  socket among multiple clients because there's no simple way to
  identify *which* client a packet should go to (we are just sending it
  into the first client with a matching Address we find).  It was also
  binding at startup, which meant even if you didn't use it, you'd still
  have a listening UDP port active.  (We'll revisit this shortly in a
  larger changeset).

- The default local bind address is now [::]:0 for clients, which is
  generally the right thing (libuv auto-magics IPv4 support when using
  it).

- Removed the block/unblock code; it is a vestigal remain of Lokinet's
  TCP tunnel blocking code, which doesn't exist in libquic.

- Mark the `server_data_callback_t` as deprecated/broken; we don't
  really need it and it was kind of a nuissance to make it work through
  the libuv udp refactor.  (Will come back to delete this later, but
  some tests, likely broken here, are still using it).

- Privatize and protect some Connection internals
    - Stream doesn't need it the udp handle
    - the internal ngtcp2_conn unique_ptr shouldn't be exposed
    - tighten up dangerous pointer implicit conversion with a template
      (to prevent evil things like implicit bool conversion through the
      pointer operator).
    - remove `operator ngtcp2_conn&()`: since this is always passed to
      ngtcp2 by pointer it seems unlikely we will ever have a use for
      this.

- Add src/internal.hpp for internal-only common code (i.e. that
  shouldn't be exposed in the public headers).

jagerman committed:
Use path MTU discovery

jagerman committed:
Remove obsolete ternary

ts is always set.

jagerman committed:
Log correction

It's a connection that gets congested, not a stream.

jagerman committed:
Unload the shotgun

`io_result` had a bad bug in it by being *implicitly* convertible to
bool, but that meant code such as:

    auto rv = /*something returning io_result*/;
    if (rv != 0) {
        // Error!
    }

was incredibly unintuitively doing the exact opposite thing you would
expect, because the only comparison the compiler can find there is with
the implicit conversion to bool, thus it becomes:

    if ((bool) rv != (bool) 0)

i.e.

    if ((rv.error_code == 0) != false)

i.e.

    if ((rv.error_code == 0) == true)

i.e.

    if (rv.error_code == 0)

This commit unloads the shotgun by adding `.success()` and `.failure()`
methods that are more descriptive, and removing the `operator bool()`
completely (although it would be fine if explicit, `if (rv.success())`
and `if (rv.failure())` just look nicer than `if (rv)` and `if (!rv)`,
especially when working with C code where this would be inverted from
logic on the raw int return values.

This logic inversion also massively improves a performance bottleneck
where we were exiting the flush streams after a single packet (because
we'd interpret success as a failure).

jagerman committed:
Simplification

This loop doesn't throw (aside from catastrophic failures like memory
allocation failure, but we can't really do anything useful if that
happens anyway).

jagerman committed:
Move "common"/"internal" targets to top-level cmake

Also removes the global `include_directories` in favour of
target-specific includes which is generally the preferred cmake
approach.

Before the commit the common/internal targets were constructed in src's
CMakeLists, but that is weird because src has to reach outside its own
directory to find include paths, and then `tests` are reaching into a
sibling for the targets they tests need.

The ../ also causes issues with some log statements because, depending
on how a header gets included, might end up being something like
"src/../include/whatever" in the filename which doesn't get stripped
because of the relative path backtracking.

jagerman committed:
Add user-defined literals for k/M/G/T and ki/Mi/etc.

The "i" ones are base-1024 values, the non-i ones are base-1000.  For
example:

    constexpr size_t ENOUGH_RAM_FOR_ANYONE = 640ki;

jagerman committed:
Make -DSUBMODULE_CHECK=OFF work

jagerman committed:
De-static global constants

jagerman committed:
Update to latest ngtcp2 `main` branch

dr7ana committed:
Server stream creation functionality added

- stream smart ptrs can be extracted via callback when ready for
  broadcast by both client and server
- servers can natively open streams and send data down them
- 006: server stream extraction
- 007: server stream creation, client/server stream extraction

dr7ana commented on include/quic/connection.hpp:
```diff
@@ -109,11 +109,9 @@ namespace oxen::quic
-        void schedule_retransmit(uint64_t ts = 0);
-
-        int init_gnutls(Client& client);
+        /// Returns a pointer to the owning endpoint; identity can be
```
was this comment a leftover from resolving a merge conflict, or did the function itself get dropped? i dont see anything in `connection.cpp`

jagerman replied:
Looks like it came in the server stream creation code.  I'll remove it.

jagerman committed:
Set default address to the IPv6 any addr

libuv takes care of making this dual-stack for both IPv4/IPv6
connectivity.

jagerman committed:
Move shutdown walk into loop call

We want this to happen before we call `stop()` to initiate an event loop
stop; currently this happens after we've stopped, which could put it
after the event loop has already stopped.

That broke things all to hell, though, because of uvw stupidity, so as a
workaround just disable the walk for now until the upcoming rework of
endpoint.

jagerman committed:
Merge conflict fixes

jagerman committed:
Remove extraneous `;` after function definitions

jagerman committed:
Remove signal handling callbacks

jagerman committed:
Nuke unwanted destructor

tewinget reviewed (approved):
LGTM
